### PR TITLE
Merkle tree update,  account DataProvider init and contracts deployment

### PR DESCRIPTION
**Pull request by shiwk**
1. add setDataProvider for account
2. contract deployment for accountZero 
3. remove concurrency in Merkle tree.

### CODE REVIEW

**Comment by shiwk** [comment]
And Genesis Block added

File: AElf.Kernel/DataProvider.cs [hunk 1]
```diff
@@ -13,21 +13,19 @@ public class DataProvider : IDataProvider
...
-        /// <param name="accountAddress"></param>
+        /// <param name="account"></param>
         public DataProvider(IAccount account)
         {
             _account = account;
 
-            keyHash = default(IHash);
-            oldValueHash = default(IHash);
-            newValueHash = default(IHash);
+            _keyHash = default(IHash);
```
Comment (loning) [inline AElf.Kernel/DataProvider.cs]: default(IHash) is null, not need...

File: AElf.Kernel/GenesisBlock.cs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    public class GenesisBlock : IBlock
+    {
+        private readonly BlockHeader _blockHeader = new BlockHeader(Hash<IBlock>.Zero);
+        private readonly BlockBody _blockBody = new BlockBody();
+        private AccountZero _accountZero;
+        private readonly WorldState _worldState;
+
+        public GenesisBlock(WorldState worldState, AccountZero accountZero)
+        {
+            _worldState = worldState;
```
Comment (loning) [inline AElf.Kernel/GenesisBlock.cs]: WorldState And AccountZero should in chain, not in a block.

GenesisBlock should have a transaction to deploy AccountZero

Comment (shiwk) [reply AElf.Kernel/GenesisBlock.cs]: I will try

File: AElf.Kernel/WorldState.cs [hunk 3]
```diff
@@ -13,11 +13,27 @@ public class WorldState : IWorldState
         private Dictionary<IAccount, IAccountDataProvider> _accountDataProviders;
 
         private BinaryMerkleTree<IHash> _merkleTree;
+        private List<IDataProvider> _dataProviders;
 
-        public WorldState()
+        private static WorldState _instance = null;
+        private static readonly object SyncObject = new object();
```
Comment (loning) [inline AElf.Kernel/WorldState.cs]: no singleton, inject from constructor

**Pull request merged**